base_layer/core/src/transaction.rs

```diff
@@ -84,10 +187,113 @@ pub struct TransactionKernel {
+    pub fn empty() -> TransactionKernel {
+        TransactionKernel { features: KernelFeatures::empty(), fee: 0, lock_height: 0, excess: None, excess_sig: None }
+    }
+
+    /// Build a transaction kernel with the provided fee
+    pub fn with_fee(mut self, fee: u64) -> TransactionKernel {
+        self.fee = fee;
+        self
+    }
+
+    /// Build a transaction kernel with the provided lock height
+    pub fn with_lock_height(mut self, lock_height: u64) -> TransactionKernel {
+        self.lock_height = lock_height;
+        self
+    }
+}
+
+/// Implement the canonical hashing function for TransactionKernel for use in ordering
+impl Hashable for TransactionKernel {
+    type Hasher = Blake256;
+
+    fn hash(&self) -> Vec<u8> {
+        let mut hasher = Self::Hasher::new();
+        hasher.input(vec![self.features.bits]);
+        let mut fee_lock_height = Vec::new();
+        fee_lock_height.write_u64::<LittleEndian>(self.fee).unwrap();
+        fee_lock_height.write_u64::<LittleEndian>(self.lock_height).unwrap();
+        hasher.input(fee_lock_height);
+        if self.excess.is_some() {
+            hasher.input(self.excess.unwrap().to_bytes());
+        }
+        if self.excess_sig.is_some() {
+            hasher.input(self.excess_sig.unwrap().get_signature().to_bytes());
+        }
+        hasher.result().to_vec()
+    }
+}
+
+hashable_ord!(TransactionKernel);
+
+/// A transaction which consists of a kernel offset and an aggregate body made up of inputs, outputs and kernels.
+pub struct Transaction {
+    /// This kernel offset will be accumulated when transactions are aggregated to prevent the "subset" problem where
+    /// kernels can be linked to inputs and outputs by testing a series of subsets and see which produce valid
+    /// transactions
+    pub offset: BlindingFactor,
+    /// The constituents of a transaction which has the same structure as the body of a block.
+    pub body: AggregateBody,
+    /// reference to the Base point used in the commitments in this transaction
+    pub base: &'static Base,
+}
+
+impl Transaction {
+    /// Create a new transaction from the provided inputs, outputs, kernels and offset
+    pub fn new(
+        base: &'static Base,
+        inputs: Vec<TransactionInput>,
+        outputs: Vec<TransactionOutput>,
+        kernels: Vec<TransactionKernel>,
+        offset: BlindingFactor,
+    ) -> Transaction
+    {
+        Transaction { base, offset, body: AggregateBody::new(inputs, outputs, kernels) }
+    }
+
+    /// Calculate the sum of the inputs and outputs including the fees
+    pub fn sum_commitments(&self, fees: u64) -> Commitment {
+        let fee_commitment = Commitment::new(&Scalar::zero(), &Scalar::from(fees), self.base);
+
+        let outputs_minus_inputs =
+            &self.body.outputs.iter().fold(Commitment::zero(self.base), |acc, val| &acc + &val.commitment) -
+                &self.body.inputs.iter().fold(Commitment::zero(self.base), |acc, val| &acc + &val.commitment);
```

Comment:
❤️ the Rust functional stuff. 
Maybe an alias / const to `Commitment::zero(self.base)` cleans this up a bit? (Or whatever we end up changing this to)

Reply:
So it can't be a const because it will be different for every `Transaction` as we have been aiming to support multiple bases. I think this will be dealt with as we refactor the `HomomorphicCommitment`